### ISSUE TITLE
401 Unauthorized error connecting to Unicorn

### ISSUE THREAD

**Issue by vasiliyfomichev** [BerserkerDotNet/Unicorn.VisualStudio#7]
The extension is installed on VS2015 Community Edition, trying to connect to Sitecore 8.1 Update 1 with WFFM module running the Habitat. Getting "Error while executing operation on server 'Habitat': response status code does not indicate success: 401 (unauthorized)." 

Unicorn 3.1 came with the Habitat.

**Comment by BerserkerDotNet** [comment]
Interesting, are you able to do a sync from Unicorn's web based control panel?

Can you check that:
1. You have correct endPoint in settings. (By default VS control panel uses /unicornRemote.aspx which is not correct, for Unicorn 3.1 it has to be /unicorn.aspx)
2. Either "Debug" is set in a Web config or you supplied a correct "Shared Secret/Security Token" in the connection info. (has to be the same as in Unicorn.UI.config).

Thanks.

**Comment by vasiliyfomichev** [comment]
It was indeed the debug="false" on the compilation node in web.config that was causing the 401. After changing it, i was able to see the configurations, however, got a new error "Error while executing operation on server 'Habitat': Index was outside the bounds of the array.' After playing around with settings a bit i noticed that the error only came up when the "Check for configuration health" was checked in settings. Any idea on what could be causing that?

**Comment by BerserkerDotNet** [comment]
"Configuration Health" option is there to warn you in the following situations:
1. Configuration does not have any serialized items.
2. Configuration does not have a valid root item
3. Configuration has some dependent configurations
4. Configuration has transparent sync enabled.

The error you are receiving means that the response from server did not contain all the necessary bits for the control panel to parse.
Can you send me a response to the "ConfigurationHealth" request, please?
Also, I will place more validation in the response handler to prevent this from happening in future.
Meanwhile, you can safely turn it off in settings.

Have you tried running "sync" operation so far?

**Comment by vasiliyfomichev** [comment]
Yep, tried the sync and it seemed to run fine without errors, although, I haven't checked to make sure it worked correctly. Here are the  [fiddler sessions](https://www.cmsbestpractices.com/Unicorn out of bounds fiddler sessions.zip) for the requests with the config health feature enabled.

**Comment by BerserkerDotNet** [comment]
That's great, if there were no errors reported during sync, then it should be fine.

Thank you very much for the fiddler session.
It appears that you are using Unicorn 3.1.0-beta09 (X-Unicorn-Version: 3.1.0-beta09), unfortunately,  "ConfigurationHealth" check was introduced recently in a stable version. 
I would recommend you to upgrade your Unicorn to the latest stable version(https://www.nuget.org/packages/Unicorn/3.1.0).

Strangely enough, Unicorn does not return a 404, but instead returns 200 OK and gives an HTML for the admin panel, as if I would request /unicorn.aspx.
I will look at the Unicorn source code to see how I can prevent this.

**Comment by kamsar** [comment]
iirc the default verb is considered to be the control panel but it would make sense to 404 on unexpected verbs, yes.

**Comment by mgurevich69** [comment]
I am having similar connection issue. Not sure if need to enter new issue or continue on this one.
Environment is Sitecore 8.1 Update 1,  Unicorn Version 3.1.7 works fine in the web browser. Debug in web.config is set to true. The latest VS plugin however has connection issue. When I leave the token out or put it in always get 404. Please help.

This is from  the Fiddler Sessions:

```

GET http://habitat.dev.local/unicorn.aspx?verb=Version HTTP/1.1
X-MC-MAC: SMLTRFA+2NsRlFC8IzJgG4bbZh+GdszXhhL1b2vrcwCwgm1uu4qJA3W7QwK9w+JeVwiHcwV7ek545I2j7kDEUQ==
X-MC-Nonce: 500fc22c8e2f4927b31b8f8f605cabd7
Host: habitat.dev.local


HTTP/1.1 404 Not Found
Content-Type: text/plain; charset=utf-8
Server: Microsoft-IIS/10.0
X-Unicorn-Version: 3.1.7
X-Powered-By: ASP.NET
Date: Tue, 02 Aug 2016 15:21:28 GMT
Content-Length: 9
Not Found
------------------------------------------------------------------

GET http://habitat.dev.local/unicorn.aspx?verb=Challenge HTTP/1.1
Host: habitat.dev.local
HTTP/1.1 200 OK



```
